Lab4/src/ca/mcgill/ecse211/lab4/UltrasonicLocalizer.java

```diff
@@ -0,0 +1,253 @@
+      odometer.setPosition(new double [] {0.0, 0.0, 0}, new boolean [] {true, true, true});
+      //We will do rising edge
+    } else {
+      //To make sure that as the sensor starts it doesn't sees that as a rising edge
+      lastDistance = 255;
+      //Set the speed of the motors
+      Nav.setSpeeds(SPEED);
+      //While there is no wall turn to the right
+      while (getFilter(SMALL_DISTANCE)>=WALL_DISTANCE - MARGIN_DISTANCE) {
+        Nav.turnClockwise();
+      }
+      //Record the first angle
+      while (true) {
+        Nav.turnClockwise();
+        //Before noise
+        if (!noiseZone && getFilter(SMALL_DISTANCE) >= WALL_DISTANCE - MARGIN_DISTANCE ) {
+          firstangle = odometer.getTheta();
+          //we have now entered the noise margin
+          noiseZone = true;
+          //In the noise margin, continue
+        } else if ( getFilter(SMALL_DISTANCE) <= WALL_DISTANCE + MARGIN_DISTANCE && noiseZone) {
+          firstangle2 = odometer.getTheta();
+          //We have exited the noise margin
+        } else if(noiseZone) {
+          noiseZone = false;
+          //Exit the loop
+          break;
+        }
+      }
+      Sound.beep();
+      //Average the angle before and after the noise margin
+      if(firstangle2!=0) {
+        firstangle = (firstangle+firstangle2)/2.0;
+      } else if(firstangle == 0) {
+        firstangle = odometer.getTheta();
+      }
+      //Change direction and wait there is a wall
+      while (getFilter(SMALL_DISTANCE)>=WALL_DISTANCE - MARGIN_DISTANCE) {
+        Nav.turnAnticlockwise();
+      }
+      // Same as for the first angle, record the angle at the edge of the wall
+      lastDistance = 0;
+      while (true) {
+        Nav.turnAnticlockwise();
+        if (!noiseZone && getFilter(SMALL_DISTANCE) >= WALL_DISTANCE - MARGIN_DISTANCE) {
+          secondangle = odometer.getTheta();
+          noiseZone = true;
+        } else if ( getFilter(SMALL_DISTANCE) <= WALL_DISTANCE + MARGIN_DISTANCE && noiseZone){
+          secondangle2 = odometer.getTheta();
+        }else if(noiseZone) {
+          noiseZone = false;
+          break;
+        }
+      }
+      Sound.beep();
+      if(secondangle2!=0) {
+        secondangle = (secondangle+secondangle2)/2.0;
+      } else if(secondangle == 0) {
+        secondangle = odometer.getTheta();
+      }
+      //stop both motors
+      Nav.stopMotors();
+      //If the first angle is bigger, wrap it around so the average 
+      //of both angle is 45 degrees
+      if(firstangle > secondangle) {
+        firstangle -= 360;
+      }
+      //calculate the average angle
+      double averageAngle = (firstangle + secondangle)/2.0;
+      double zeroangle =  averageAngle - 45 - odometer.getTheta();
+      //Turn to zero degrees
+      Nav.turnTo(zeroangle);
+      //reset the odometer to 0
+      odometer.setPosition(new double [] {0.0, 0.0, 0}, new boolean [] {true, true, true});

```

Comment:
Again, a lot of your code is basically a copy/paste with minor variations here and there. Refactor!